File: notes/src/data/mappers/NoteMapper.ts
import { NoteDTO, NoteRequestDTO } from '../contracts/NoteContract';
import { NoteModel } from '../models/NoteModel';

export function fromResponse(noteDTO: NoteDTO): NoteModel {
  if (!noteDTO) {
    return new NoteModel();
  }

  const { categories, content, creationDate, id, markdown, title } = noteDTO;

  return new NoteModel(id, new Date(creationDate), title, markdown, content, categories);
}

export function toResponse(note: NoteModel): NoteRequestDTO {
  const { id, categories, content, title, markdown } = note;
  return {
    id,
    title,
    markdown,
    content,
    categories
  };
}

File: backend/Mappers/NoteMapper.cs
using System;
using System.Linq;
using Z01.Models;
using Z4.dto;

namespace Z4.Mappers
{
    public class NoteMapper
    {
        public Note mapResponseToNote(NoteRequestDTO request)
        {
            return new Note
            {
                Description = request.Description,
                Markdown = request.Markdown,
                NoteID = request.NoteID ?? 0,
                Title = request.Title,
                RowVersion = request.RowVersion,
                NoteDate = DateTime.Now
            };
        }

        public NoteResponseDTO mapNoteToResponse(Note note)
        {
            return new NoteResponseDTO
            {
                Description = note.Description ?? "",
                Markdown = note.Markdown,
                NoteID = note.NoteID,
                Title = note.Title,
                RowVersion = note.RowVersion,
                NoteDate = note.NoteDate,
                Categories = note.NoteCategories.Select(_ => _.Category.Name)
            };
        }
    }
}
